.github/workflows/reusable-mypy.yml

```diff
@@ -0,0 +1,21 @@
+on:
+  workflow_call:
+
+jobs:
+  check-with-mypy:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v4
+
+      - name: Install Python
+        uses: actions/setup-python@v5
+        with:
+          python-version: 3.x
+
+      - name: Install dependencies and type information
+        run: |
+          python -m pip install --upgrade pip
+          python -m pip install mypy
+
+      - name: mypy static type check
+        run: mkdir .mypy_cache && mypy --install-types --ignore-missing-imports --non-interactive --pretty .
```

Comment:
Skimming the available `mypy` options, there were just a few optional warnings (see the [docs](https://mypy.readthedocs.io/en/stable/command_line.html#configuring-warnings)) that caught my eye that we may want to consider enabling:
- `--warn-redundant-casts`
- `--warn-unused-ignores`
- `--warn-unreachable`
- `--strict-equality`

Reply:
I also don't know if the `--warn-*` flags will cause a non-zero exit code if they do catch any problems, I think ideally we *would* want the action to fail in that case, but not sure of the actual behavior.